File: backend/app/views/birthday_settings.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field, field_validator


class BirthdayChannel(StrEnum):
    SMS = "SMS"
    EMAIL = "EMAIL"
    BOTH = "BOTH"


class BirthdaySettingsData(BaseModel):
    notify_days_before: int
    enabled: bool
    channel: BirthdayChannel
    send_hour: int
    send_minute: int
    company_name: str
    message_template: str
    email_subject: str
    email_message_template: str
    auto_assign_gift: bool
    last_run_at: datetime | None


class BirthdaySettingsResponse(BaseModel):
    success: bool = True
    data: BirthdaySettingsData
    meta: dict = {}


class BirthdaySettingsUpdate(BaseModel):
    notify_days_before: int = Field(ge=0, le=30)
    enabled: bool
    channel: BirthdayChannel
    send_hour: int = Field(ge=0, le=23)
    send_minute: int = Field(ge=0, le=59)
    company_name: str = Field(max_length=120)
    message_template: str = Field(min_length=1, max_length=500)
    email_subject: str = Field(min_length=1, max_length=255)
    email_message_template: str = Field(min_length=1, max_length=2000)
    auto_assign_gift: bool

    @field_validator("message_template", "email_subject", "email_message_template")
    @classmethod
    def _not_blank(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("This field cannot be blank")
        return stripped

    @field_validator("company_name")
    @classmethod
    def _strip_company_name(cls, value: str) -> str:
        return value.strip()

File: frontend/lib/api/birthday-settings.ts
import { apiGet, apiPut } from "@/lib/api/client";
import type { ApiEnvelope } from "@/lib/api/types";

export type BirthdayChannel = "SMS" | "EMAIL" | "BOTH";

export interface BirthdaySettings {
  notifyDaysBefore: number;
  enabled: boolean;
  channel: BirthdayChannel;
  sendHour: number;
  sendMinute: number;
  companyName: string;
  messageTemplate: string;
  emailSubject: string;
  emailMessageTemplate: string;
  autoAssignGift: boolean;
  lastRunAt: string | null;
}

export async function getBirthdaySettings(): Promise<BirthdaySettings> {
  const envelope = await apiGet<ApiEnvelope<BirthdaySettings>>("/settings/birthday");
  return envelope.data;
}

export async function updateBirthdaySettings(
  input: Omit<BirthdaySettings, "lastRunAt">
): Promise<BirthdaySettings> {
  const envelope = await apiPut<ApiEnvelope<BirthdaySettings>>("/settings/birthday", {
    notify_days_before: input.notifyDaysBefore,
    enabled: input.enabled,
    channel: input.channel,
    send_hour: input.sendHour,
    send_minute: input.sendMinute,
    company_name: input.companyName,
    message_template: input.messageTemplate,
    email_subject: input.emailSubject,
    email_message_template: input.emailMessageTemplate,
    auto_assign_gift: input.autoAssignGift,
  });
  return envelope.data;
}
